Controller/Adminhtml/IndexingManager/Reindex.php

```diff
@@ -30,34 +36,80 @@ public function __construct(
     /**
      * @throws ExceededRetriesException
      * @throws AlgoliaException
-     * @throws NoSuchEntityException
+     * @throws NoSuchEntityException|DiagnosticsException
      */
     public function execute()
     {
-        $storeIds = $this->getRequest()->getParam("store_id") === '0' ?
+        $params = $this->getRequest()->getParams();
+        $storeIds = !isset($params["store_id"]) || $params["store_id"] === '0' ?
             array_keys($this->storeManager->getStores()) :
-            [(int) $this->getRequest()->getParam("store_id")];
+            [(int) $params["store_id"]];
 
-        $entities = $this->getRequest()->getParam("entity") === 'all' ?
-            ['products', 'categories', 'pages'] :
-            [$this->getRequest()->getParam("entity")];
+        $entities = $this->defineEntitiesToIndex($params);
+        $entityIds = $params['selected'] ?? null;
 
-        $this->reindexEntities($entities, $storeIds);
+        $this->reindexEntities($entities, $storeIds, $entityIds);
 
         /** @var Redirect $resultRedirect */
         $resultRedirect = $this->resultFactory->create(ResultFactory::TYPE_REDIRECT);
 
-        return $resultRedirect->setPath('*/*/');
+        return $resultRedirect->setPath($this->defineRedirectPath($params));
+    }
+
+    /**
+     * @return array|string[]
+     */
+    protected function defineEntitiesToIndex(array $params): array
+    {
+        $entities = [];
+        if (isset($params["entity"])) {
```

Comment:
Good we self document by defining boolean functions to evaluate the params like `isMassAction` or `isFullIndex`?

Reply:
True, this part wasn't very clear I agree